Add endogenous transport option

s8au opened the pull request:
- select with endogenous_transport in config file
- it is possible to select transport shares for part of the fuel types
- creates ICE-vehicle-fleet for baseyear if myopic is selected and no transport share is defined
- adds constraints for endogenously chosen EVs in solve_network

relies on PR in technology-data repository: [BertoGBG:master](https://github.com/PyPSA/technology-data/pull/106)
Closes https://github.com/PyPSA/pypsa-eur/issues/507.

## Changes proposed in this Pull Request


## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv`.
- [x] A release note `doc/release_notes.rst` is added.

martavp reviewed (commented):
Many thanks @s8au , this looks great! 

I still need to run the new code and I'll do it soon, but in the meantime see some suggestions based on reading the changes.

I think you can check "Changed dependencies are added to envs/environment.yaml." since no changes are required. 

Try also to add a release note to doc/release_notes.rst

martavp commented:
The current definition of p_nom for the link "land transport EV" needs to be changed. 
Currently, it is based on the maximum value of the transport demand 

> n.links.loc[
>                    (n.links.carrier == "land transport EV")
>                    & (n.links.bus1.str.contains(place)),
>                   "p_nom",
>                ] = electric_share * max(p_nom[place])


This represents the minimum number of cars needed to supply that demand, but in reality, cars are most of the time parked and we should use the total number of cars, as it was done before

https://github.com/PyPSA/pypsa-eur/blob/706942957ba2d8a11a669357adb6d6289b5fea78/scripts/prepare_sector_network.py#L1501

martavp commented:
There is a bug when adding the storage for EVs exogenous. The following code

> n.stores.loc[
>                       (n.stores.carrier == "EV battery storage")
>                       & (n.links.bus1.str.contains(place)),
>                        "e_nom",
>                   ] = (
>                        number_cars[place]
>                        * options.get("bev_energy", 0.05)
>                        * options["bev_availability"]
>                        * electric_share
>                    )


Should be:


> n.stores.loc[
>                       (n.stores.carrier == "EV battery storage")
>                       & (n.stores.bus.str.contains(place)),
>                        "e_nom",
>                   ] = (
>                        number_cars[place]
>                        * options.get("bev_energy", 0.05)
>                        * options["bev_availability"]
>                        * electric_share
>                    )

martavp commented:
The constraint relating the BEV charge rate and the storage capacity of the batteries, needs to take into account that [in the config file it can be selected that only a percentage of the BEV provides flexibility 
](https://github.com/PyPSA/pypsa-eur/blob/706942957ba2d8a11a669357adb6d6289b5fea78/config/config.default.yaml#L364)
> lhs2 = (
>        n.model.variables["Store-e_nom"].loc[ev_store]
>        / n.config["sector"]["bev_energy"]
>    )

should be: 
> lhs2 = (
>        n.model.variables["Store-e_nom"].loc[ev_store]
>        / (n.config["sector"]["bev_energy"]*n.config["sector"]["bev_availability"])
>    )

martavp commented:
I found a bug in the definition of p_max_pu and p_min_pu of the link "land transport oil"
The following 

> n.links_t.p_max_pu[
>                   n.links[
>                        (n.links.carrier == "land transport oil")
>                        & (n.links.bus1.str.contains(place))
>                   ].index.values[0]
>                ] = ((p_nom[place]) / max(p_nom[place]) * ice_efficiency / p_set[place])

should be:
> n.links_t.p_max_pu[
>                   n.links[
>                        (n.links.carrier == "land transport oil")
>                        & (n.links.bus1.str.contains(place))
>                   ].index.values[0]
>                ] = ((p_nom[place]))
otherwise, since p_nom and p_set are defined to be equal on a previous line, p_max_pu becomes a constant number and the flow of energy through that link is not able to supply the land transport demand. 


**This should be changed for p_max_pu and p_min pu of both links 'land transport oil' and 'land transport fuel cell'**

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
-            / options["transport_fuel_cell_efficiency"]
-            * transport[nodes],
+            lifetime=costs.at["Hydrogen fuel cell (passenger cars)", "lifetime"],
+            efficiency=1.0 / options["transport_fuel_cell_efficiency"] * p_set[nodes],
```
I used the same syntax as in the p_set for the endogenous transport https://github.com/PyPSA/pypsa-eur/blob/fe961e6704a1acc7e666e755149c6cd80ec34bac/scripts/prepare_sector_network.py#L1559C12-L1559C12 , assuming that transport_fuel_cell_efficiency is not the actual efficiency but the adjustment to change from the EV efficiency to the H2 efficiency.

s8au replied:
I agree now and changed it in  239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+            p_nom_extendable=True,
         )
+        # exogenous pathway if fuel_cell_share is defined
+        if float(fuel_cell_share or 0) > 0:
```
The function float() cannot handle NoneType arguments like None, which we get if the shares are commented out for the exogenous land transport. Here, float gets passed the value 0 if fuel_cell_share is None, thus mostly preserving the previous syntax. But I am also open to other suggestions.

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
 
-    if ice_share > 0:
+    # add internal combustion engine vehicle links
+    if ice_share != 0:
```
As above, this would throw an error if no ice_share is defined, thus setting ice_share to None

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                    )
+
+    # Add hydrogen vehicle links
+    if fuel_cell_share != 0:
```
Same reasoning as with ice_share

s8au commented on config/config.default.yaml:
```diff
@@ -369,6 +369,11 @@ sector:
   v2g: true
+  endogenous_transport: false
+  energy_to_cars: 0.01 #Mw@wheel/car assuming 0.2kwh/km and 50km/h
+  EV_consumption_1car: 0.01 #MWh_elec/hour (assuming 0.2 kWh/km https://github.com/PyPSA/pypsa-eur/blob/1fbe971ab8dab60d972d3a7b905b9cec7171c0ad/config/config.default.yaml#L382 and velocity of 50km/h
```
I kept EV_consumption_1car in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on config/config.default.yaml:
```diff
@@ -369,6 +369,11 @@ sector:
+  energy_to_cars: 0.01 #Mw@wheel/car assuming 0.2kwh/km and 50km/h
+  EV_consumption_1car: 0.01 #MWh_elec/hour (assuming 0.2 kWh/km https://github.com/PyPSA/pypsa-eur/blob/1fbe971ab8dab60d972d3a7b905b9cec7171c0ad/config/config.default.yaml#L382 and velocity of 50km/h
+  ICE_consumption_1car: 0.033 #MWh_oil/hour (assuming 0.66 kWh_oil/km and 50 km/h (with the link efficiency 0.3, they are equivalent to 0.01 MWh_elec/hour)
+  H2_consumption_1car: 0.02 #MWh_H2/hour (assuming 0.4 kWh_H2/km and 50km/h) (with efficiency 0.5, they are equivalent to 0.01 MWh_elec/hour)
```
Should be now documented in the configtables csv in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+    if options["endogenous_transport"] == False:
+        if any([fuel_cell_share == None, electric_share == None, ice_share == None]):
+            logger.warning(
+                f"Exogenous transport selected, but not all transport shares are defined."
```
Fixed in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
             bus0=nodes,
-            bus1=nodes + " EV battery",
-            p_nom=p_nom,
+            bus1=nodes + " EV driving",
```
I changed it back to "EV battery" in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
             nodes,
             location=nodes,
-            suffix=" EV battery",
+            suffix=" EV driving",
```
Agreed, changed in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                nodes,
+                suffix=" V2G",
+                bus1=nodes,
+                bus0=nodes + " EV driving",
```
Agreed, changed in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                "Store",
+                nodes,
+                suffix=" EV battery storage",
+                bus=nodes + " EV driving",
```
Agreed, changed in 239bee993d928bfd9c0fcd58d185cd3055d26430

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                e_nom_extendable=True,
+                e_max_pu=1,
+                e_min_pu=dsm_profile[nodes],
+                # lifetime?
```
I set the battery lifetime to the car lifetime in commit 239bee993d928bfd9c0fcd58d185cd3055d26430 while I was changing the constraints in the solve network to prevent them being used as v2g while the land transport EVs are already decomissioned, but we can discuss this

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1571,32 +1722,53 @@ def add_land_transport(n, costs):
+                * options["transport_internal_combustion_efficiency"]
+            ),
+            efficiency=1.0
+            / options["transport_internal_combustion_efficiency"]
```
Agreed, changed in 239bee993d928bfd9c0fcd58d185cd3055d26430

lisazeyen commented:
@martavp @s8au thanks a lot for all your work on this!

 Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:

1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving.
2.)  Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.

@s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?

s8au commented:
> @martavp @s8au thanks a lot for all your work on this!
> 
> Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> 
> 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> 
> @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?

Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. 
I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

lisazeyen commented:
> @lisazeyen
> 
> > Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> > 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> > @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?
> 
> Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

lisazeyen commented:
> @lisazeyen
> 
> > Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> > 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> > @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?
> 
> Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

lisazeyen commented:
> @lisazeyen
> 
> > Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> > 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> > @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?
> 
> Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

lisazeyen commented:
> @lisazeyen
> 
> > Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> > 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> > @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?
> 
> Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

lisazeyen commented:
> @lisazeyen
> 
> > Tom and I just had a look at the code since we would like to have the endogenous land transport in the main repository quite soon. Two main points came to our mind:
> > 1.) As a first step it made make sense to fix the number of cars and not allow switching between cars by assuming a `p_min_pu=p_max_pu=profile transport demand` for all three links (EV, H2 Fuel Cell and ICEs). This should also speed up the solving. 2.) Further the transport demand profile `p_set` should not be shifted since this was only done for smoothing the EV charging. We can first just leave out the shifting completely when defining `p_set` . In a second step one could think about adding a time-dependent efficiency to the EV links to account for different charging times.
> > @s8au would you have time to include it in the next two weeks or should we give it a try? Did you have time to check also for newer cost assumptions for EVs?
> 
> Thank you both for looking over it and your suggestions. I will try to include the p_min_pu/p_max_pu constraint and remove the smoothing next week. I looked a bit around but so far I only found a datasheet for [commercial freight and passenger transport ](https://ens.dk/sites/ens.dk/files/Analyser/technology_data_for_commercial_freight_and_passenger_transport.pdf). Did you have something specific in mind?

Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+            carrier="land transport EV",
+            lifetime=costs.at["Battery electric (passenger cars)", "lifetime"],
+            capital_cost=costs.at["Battery electric (passenger cars)", "fixed"]
+            / (options["energy_to_cars"] * options.get("bev_charge_efficiency", 0.9)),
```
Addressed in commit 659f97743c204598e65afe7695cf122826d6fed9:
[1] I kept EV_consumption_1car because I find it more intuitive 
[2] I agree

martavp replied:
perfect

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                    p_nom[place]
+                    / max(p_nom[place])
+                    / costs.at["Battery electric (passenger cars)", "efficiency"]
+                )
```
In commit 659f97743c204598e65afe7695cf122826d6fed9:
[1] Yes, I changed that.
[2] I tried using a more compact form now

martavp replied:
great

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+            lifetime=costs.at["Hydrogen fuel cell (passenger cars)", "lifetime"],
+            efficiency=1.0 / options["transport_fuel_cell_efficiency"] * p_set[nodes],
+            capital_cost=costs.at["Hydrogen fuel cell (passenger cars)", "fixed"]
+            / (options["energy_to_cars"] * options["transport_fuel_cell_efficiency"]),
```
Yes, I agree, updated in 659f97743c204598e65afe7695cf122826d6fed9
I kept H2_consumption_1car

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
-            * electric_share
-        )
+        # exogenous pathway if electric share is defined
+        if float(electric_share or 0) > 0:
```
As seen below
https://github.com/PyPSA/pypsa-eur/pull/734#discussion_r1328765287

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1571,32 +1722,53 @@ def add_land_transport(n, costs):
+            capital_cost=costs.at["Liquid fuels ICE (passenger cars)", "fixed"]
+            / (
+                options["energy_to_cars"]
+                * options["transport_internal_combustion_efficiency"]
```
I kept ICE_consumption_1car in commit 2dd61a3b0cf15e1f55531edc4abddb42fd626b69

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                    / max(p_nom[place])
+                    * options["transport_fuel_cell_efficiency"]
+                    / p_set[place]
+                )
```
Should be fine now, although I think I need to keep the transport_fuel_cell_efficiency. Do you agree with how I did it in this commit: 3e4627507dcb0ba062a5c161a93204921cf9a39f ?

martavp replied:
Yes, I do. However, three comments:

(a) the re-definition of p_max_pu and p_min_pu if '  if float(ice_share or 0) > 0:' is commented out (so it is not taking place)

(b) remove all the commented lines to clean the code

(c) The definition 'p_nom = p_set' for 'float(fuel_cell_share or 0) > 0:' is confusing. Why don't just use p_set

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1571,32 +1722,53 @@ def add_land_transport(n, costs):
+                        (n.links.carrier == "land transport oil")
+                        & (n.links.bus1.str.contains(place))
+                    ].index.values[0]
+                ] = ((p_nom[place]) / max(p_nom[place]) * ice_efficiency / p_set[place])
```
Also here, check out in commit 3e4627507dcb0ba062a5c161a93204921cf9a39f if you agree with my code

s8au commented on scripts/add_existing_baseyear.py:
```diff
@@ -606,6 +606,43 @@ def add_heating_capacities_installed_before_baseyear(
+            bus2="co2 atmosphere",
+            carrier="land transport oil",
+            capital_cost=0,
+            efficiency=1.0 / ice_efficiency * p_set_year[nodes],
```
should be fixed now in commit 3e4627507dcb0ba062a5c161a93204921cf9a39f

s8au commented on scripts/add_existing_baseyear.py:
```diff
@@ -606,6 +606,43 @@ def add_heating_capacities_installed_before_baseyear(
+            efficiency2=costs.at["oil", "CO2 intensity"],
+            lifetime=costs.at["Liquid fuels ICE (passenger cars)", "lifetime"],
+            p_nom=set_p_nom,
+            p_min_pu=p_set_year / set_p_nom * ice_efficiency / p_set_year[nodes],
```
I updated it, see commit 3e4627507dcb0ba062a5c161a93204921cf9a39f

s8au commented on scripts/prepare_sector_network.py:
```diff
@@ -1455,113 +1455,264 @@ def add_land_transport(n, costs):
+                    (n.links.carrier == "land transport EV")
+                    & (n.links.bus1.str.contains(place)),
+                    "p_nom",
+                ] = electric_share * max(p_nom[place])
```
I changed the p_nom for the exogenous EVs in commit 3e4627507dcb0ba062a5c161a93204921cf9a39f to have all the cars available

martavp commented:
> Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!

BNEF include cost assumption for trucks in Figure 4 in 
https://assets.bbhub.io/professional/sites/24/2431510_BNEFElectricVehicleOutlook2023_ExecSummary.pdf
(here, we agree that using DEA is better)

and cost assumptions for EVs until 2030 in Figure 28. https://www.transportenvironment.org/discover/hitting-the-ev-inflection-point/
(because the data only covers until 2030 and the report includes a disclaimer saying that we should not use the data, I suggest we keep using Fraunhofer data), unless @s8au and I have missed some additional BNEF database?

But you are totally right @lisazeyen that is too pessimistic about EV costs. See below the cost comparison for EVs between our current assumptions and BNEF (red dots).  We will implement a cost sensitivity analysis soon so that we can learn about the impacts. 

![BNEF_vs_Fraunhofer](https://github.com/PyPSA/pypsa-eur/assets/30744159/2b255394-05ec-4440-ab9f-3ed49c225042)

lisazeyen commented:
> > Great! I think the technology data is a good source for the heavy duty vehicles. For the light vehicles I guess there is a Bloomberg report for at least EVs. But I think every source for light vehicles which is a bit more up to date (like after 2020) should be fine!
> 
> BNEF include cost assumption for trucks in Figure 4 in https://assets.bbhub.io/professional/sites/24/2431510_BNEFElectricVehicleOutlook2023_ExecSummary.pdf (here, we agree that using DEA is better)
> 
> and cost assumptions for EVs until 2030 in Figure 28. https://www.transportenvironment.org/discover/hitting-the-ev-inflection-point/ (because the data only covers until 2030 and the report includes a disclaimer saying that we should not use the data, I suggest we keep using Fraunhofer data), unless @s8au and I have missed some additional BNEF database?
> 
> But you are totally right @lisazeyen that is too pessimistic about EV costs. See below the cost comparison for EVs between our current assumptions and BNEF (red dots). We will implement a cost sensitivity analysis soon so that we can learn about the impacts.
> 
> ![BNEF_vs_Fraunhofer](https://user-images.githubusercontent.com/30744159/279953488-2b255394-05ec-4440-ab9f-3ed49c225042.jpg)

Yes, I guess a cost sensitivity is the best way!

s8au commented:
> The current definition of p_nom for the link "land transport EV" needs to be changed. Currently, it is based on the maximum value of the transport demand
> 
> > n.links.loc[
> > (n.links.carrier == "land transport EV")
> > & (n.links.bus1.str.contains(place)),
> > "p_nom",
> > ] = electric_share * max(p_nom[place])
> 
> This represents the minimum number of cars needed to supply that demand, but in reality, cars are most of the time parked and we should use the total number of cars, as it was done before
> 
> https://github.com/PyPSA/pypsa-eur/blob/706942957ba2d8a11a669357adb6d6289b5fea78/scripts/prepare_sector_network.py#L1501
> 
> This is important to avoid problems when calculating the cost associated with cars.

I updated it for the exogenous land transport EV 
I also indicate with the :rocket: that I fixed the problems in the other comments

martavp commented on scripts/prepare_sector_network.py:
```diff
@@ -1479,89 +1514,147 @@ def add_land_transport(n, costs):
+            # set p_nom to fulfill electric share, capital cost to 0
+            n.links.loc[n.links.carrier=='land transport EV', "p_nom_extendable"] = False
+
+            p_nom = number_cars * options.get("bev_charge_rate", 0.011) * electric_share #electric_share * p_set.max(axis=0)
```
This link represents the "capacity" of EVs to supply transport demand.
However, the current definition represents the capacity for chargin the cars 
sinc 'options.get("bev_charge_rate", 0.011)'

martavp commented:
Hi @s8au ,

I mentioned to @lisazeyen yesterday that she can start reviewing this. 
To facilitate the review, Can you update the PR to the last version of the master branch and check that everything run smoothly? This facilitates the revision (currently endogenous-transport-related and non-related changes are shown and make the revision cumbersome)

There are two additional small issues:

[ ] 1. I think the link representing the "capacity" of EVs to supply transport demand is not correct.
The current definition represents the capacity for charging the cars 
since `options.get("bev_charge_rate", 0.011)` is the capacity to charge a car. 
Can you please correct this? (similarly to the exogenous capacity definition for H2 and ICE cars)?
[] 2. I understand how you estimated the capacity of the "land transport oil." In essence, you calculate the minimum capacity needed to supply the demand for land transport today and then decommission a percentage of it in every future planning horizon. The approach is fine for me. 

However, in reality, we have many more cars than "needed". Can you calculate the ratio between the existing number of cars and the cars that you calculate that are needed to supply the demand and write it as a log message? 
Something like:
 _"Although, existing ICE vehicles in 2020 account for XXXX, the model assumes only the minimum number of ICE vehicles to supply the transport demand which represents XXX"_

s8au commented:
@martavp @lisazeyen I updated the branch with the master branch and I ran a test case for the exogenous and the endogenous transport each (resulting in again being behind by 4 commits). Since these are only minor changes I would prefer to update them with additional changes to my code.

s8au commented:
I updated the PR with the newest version, so we can discuss the changes, but I still have to check if the added code in the add_brownfield works because the Gurobi license on our cluster has expired. 
@lisazeyen 
I had to move the calculation of the profiles for the p_min_pu/p_max_pu constraints after the links have their new temporal resolution because the efficiencies are now time dependent with the temperature corrections. Due to this, taking the mean of the profiles gives wrong values. 
Gurobi has some trouble with setting these profiles and then solving the land transport when the links p_nom_extendable is set to True. I am currently trying to only constrain two out of three links directly (which works fine) and then constrain the third link for the myopic pathway in the add_brownfield script (WIP).
The exogenous pathway should work fine, it runs - depending on the CO2 budget - until 2040/2050 before becoming infeasible. But that was also what I experienced with the unmodified code from the repository.

s8au commented:
@martavp I added a commit with the fixed bug, so it should run now

lisazeyen commented:
@s8au @martavp I am currently planning to integrate as the land transport demand as one single node and then the links for the different car types.  I made a suggestion, which you can see the PR here #957. I think in you current implementation you can fix the profiles when adding but you need to be adjusted the profiles depending on the efficiencies.  The only thing which is missing in the PR is the cycling shift of the EVs charging.

fneum commented:
superceded by #960
